EIPS/eip-web3-access.md

```diff
@@ -0,0 +1,109 @@
+---
+eip: <to be assigned>
+title: Opt-in web3 access
+author: Paul Bouchon <mail@bitpshr.net>
+discussions-to: https://ethereum-magicians.org/t/opt-in-web3-access/414
+status: Draft
+type: Standards Track
+category: Interface
+created: 2018-05-04
+---
+
+## Simple summary
+
+This proposal describes a new way for environments to expose the web3 API that requires user approval.
+
+## Abstract
+
+MetaMask and most other tools that provide access to web3-enabled environments do so automatically and without user consent. This exposes users of such environments to fingerprinting attacks since untrusted websites can check for a `web3` object and reliably identify web3-enabled clients.
+
+This proposal outlines a new dapp initialization strategy in which websites request access to the web3 API instead of relying on its preexistence in a given environment.
+
+## Specification
+
+### Typical dapp initialization
+
+```
+START dapp
+IF web3 is defined
+    CONTINUE dapp
+IF web3 is undefined
+    STOP dapp
+```
+
+### Proposed dapp initialization
+
+```
+START dapp
+IF web3 is defined
+    CONTINUE dapp
+IF web3 is undefined
+    REQUEST[1] web3 API
+    IF user approves
+        INJECT[2] web3 API
+        NOTIFY[3] dapp
+        CONTINUE dapp
+    IF user rejects
+        STOP dapp
+```
+
+**REQUEST[1]** This operation would be achieved by an implementation-level messaging API like `window.postMessage` and should pass a payload containing a `type` property with a value of “WEB3_API_REQUEST”.
```

Comment:
Is there a way to do this that works across all major browsers?  Ideally something that has been standardized.

Reply:
While this is just a suggestion as this proposal aims to stay platform-agnostic, `window.postMessage` works across [all major browsers](https://caniuse.com/#search=postMessage) and browser runtimes that most dapp browsers are built on. It's part of the [HTML living standard](https://html.spec.whatwg.org/multipage/comms.html#messageevent), so unless a particular environment chooses to intentionally omit a part of the HTML specification, this should be safe to use.